src/mpb.cpp

```diff
@@ -596,22 +596,21 @@ void *fields::get_eigenmode(double frequency, direction d, const volume where, c
     scalar_complex p = {real(dp->get_p()),imag(dp->get_p())};
 
     // compute sum of (kparallel+G)^2 in all the periodic directions
-    double k2sum = 0;
+    double k2sum = 0, ktmp = 0;
+    int m = 0;
     LOOP_OVER_DIRECTIONS(v.dim, dd) {
-      double ktmp = 0;
-      int m = 0;
-      if ((float(eig_vol.in_direction(dd)) == float(v.in_direction(dd))) &&
-          (boundaries[High][dd] == Periodic && boundaries[Low][dd] == Periodic)) {
-        m = dp->get_g()[dd - X];
-        ktmp = kpoint.in_direction(dd) + m/v.in_direction(dd);
-        k2sum += ktmp*ktmp;
-      }
+      m = dp->get_g()[dd - X];
+      ktmp = kpoint.in_direction(dd);
+      if ((m != 0) &&
+          (boundaries[High][dd] == Periodic && boundaries[Low][dd] == Periodic))
+        ktmp += m/eig_vol.in_direction(dd);
+      k2sum += ktmp*ktmp;
     }
 
     // compute kperp (if it is non evanescent) OR
     // frequency from kperp^2 and sum of (kparallel+G)^2
     LOOP_OVER_DIRECTIONS(v.dim, dd) {
-      if ((float(eig_vol.in_direction(dd)) != float(v.in_direction(dd))) &&
+      if ((float(eig_vol.in_direction(dd)) == 0) &&
           (boundaries[High][dd] == Periodic && boundaries[Low][dd] == Periodic)) {
```

Comment:
Probably would be better to call `normal_direction` to get the perpendicular direction?

Reply:
i.e. `eig_vol.normal_direction()`.

Reply:
(also the cell doesn't need to be periodic in that direction)